lib/srv/db/opensearch/engine.go

```diff
@@ -0,0 +1,331 @@
+		Session:               awsSession,
+		GetSigningCredentials: e.GetSigningCredsFn,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	e.signingSvc = svc
+
+	return nil
+}
+
+// ErrorDetails contains error details.
+type ErrorDetails struct {
+	Reason string `json:"reason"`
+	Type   string `json:"type"`
+}
+
+// ErrorResponse will be returned to the client in case of error.
+type ErrorResponse struct {
+	Error  ErrorDetails `json:"error"`
+	Status int          `json:"status"`
+}
+
+// SendError sends an error to OpenSearch client.
+func (e *Engine) SendError(err error) {
+	if e.clientConn == nil || err == nil || utils.IsOKNetworkError(err) {
+		return
+	}
+
+	cause := ErrorResponse{
+		Error: ErrorDetails{
+			Reason: err.Error(),
+			Type:   "internal_server_error_exception",
+		},
+		Status: http.StatusInternalServerError,
+	}
+
+	// Different error for access denied case.
+	if trace.IsAccessDenied(err) {
+		cause.Status = http.StatusUnauthorized
+		cause.Error.Type = "access_denied_exception"
+	}
+
+	jsonBody, err := json.Marshal(cause)
+	if err != nil {
+		e.Log.WithError(err).Error("failed to marshal error response")
+		return
+	}
+
+	response := &http.Response{
+		ProtoMajor: 1,
+		ProtoMinor: 1,
+		StatusCode: cause.Status,
+		Body:       io.NopCloser(bytes.NewBuffer(jsonBody)),
+		Header: map[string][]string{
+			"Content-Type":   {"application/json"},
+			"Content-Length": {strconv.Itoa(len(jsonBody))},
+		},
+	}
+
+	if err := response.Write(e.clientConn); err != nil {
+		e.Log.WithError(err).Errorf("OpenSearch: failed to send an error to the client.")
+		return
+	}
+}
+
+// HandleConnection authorizes the incoming client connection, connects to the
+// target OpenSearch server and starts proxying requests between client/server.
+func (e *Engine) HandleConnection(ctx context.Context, _ *common.Session) error {
+	err := e.checkAccess(ctx)
+	e.Audit.OnSessionStart(e.Context, e.sessionCtx, err)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer e.Audit.OnSessionEnd(e.Context, e.sessionCtx)
+
+	// TODO(Tener):
+	//  Consider rewriting to support HTTP2 clients.
+	//  Ideally we should have shared middleware for DB clients using HTTP, instead of separate per-engine implementations.
+
+	clientConnReader := bufio.NewReader(e.clientConn)
+	for {
+		req, err := http.ReadRequest(clientConnReader)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		if err := e.process(ctx, req); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+}
+
+// process reads request from connected OpenSearch client, processes the requests/responses and send data back
+// to the client.
+func (e *Engine) process(ctx context.Context, req *http.Request) error {
+	reqCopy, err := e.rewriteRequest(ctx, req)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	e.emitAuditEvent(reqCopy)
+
+	signedReq, err := e.getSignedRequest(reqCopy)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	tr, err := e.getTransport(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
```

Comment:
nit: in theory the main loop can process many requests:
```go
for {
		req, err := http.ReadRequest(clientConnReader)
		if err != nil {
			return trace.Wrap(err)
		}

		if err := e.process(ctx, req); err != nil {
			return trace.Wrap(err)
		}
	}
```
Can we create transport only once per connection not per requests ?

Reply:
Done.